File: frontend/cadastros/lib/app/data/model/cliente_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cadastros/app/infra/infra_imports.dart';
import 'package:cadastros/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class ClienteModel {
	int? id;
	int? idPessoa;
	int? idTabelaPreco;
	DateTime? desde;
	DateTime? dataCadastro;
	double? taxaDesconto;
	double? limiteCredito;
	String? observacao;
	TabelaPrecoModel? tabelaPrecoModel;

	ClienteModel({
		this.id,
		this.idPessoa,
		this.idTabelaPreco,
		this.desde,
		this.dataCadastro,
		this.taxaDesconto,
		this.limiteCredito,
		this.observacao,
		this.tabelaPrecoModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'desde',
		'data_cadastro',
		'taxa_desconto',
		'limite_credito',
		'observacao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Desde',
		'Data Cadastro',
		'Taxa Desconto',
		'Limite Credito',
		'Observacao',
	];

	ClienteModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idPessoa = jsonData['idPessoa'];
		idTabelaPreco = jsonData['idTabelaPreco'];
		desde = jsonData['desde'] != null ? DateTime.tryParse(jsonData['desde']) : null;
		dataCadastro = jsonData['dataCadastro'] != null ? DateTime.tryParse(jsonData['dataCadastro']) : null;
		taxaDesconto = jsonData['taxaDesconto']?.toDouble();
		limiteCredito = jsonData['limiteCredito']?.toDouble();
		observacao = jsonData['observacao'];
		tabelaPrecoModel = jsonData['tabelaPrecoModel'] == null ? TabelaPrecoModel() : TabelaPrecoModel.fromJson(jsonData['tabelaPrecoModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idPessoa'] = idPessoa != 0 ? idPessoa : null;
		jsonData['idTabelaPreco'] = idTabelaPreco != 0 ? idTabelaPreco : null;
		jsonData['desde'] = desde != null ? DateFormat('yyyy-MM-ddT00:00:00').format(desde!) : null;
		jsonData['dataCadastro'] = dataCadastro != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataCadastro!) : null;
		jsonData['taxaDesconto'] = taxaDesconto;
		jsonData['limiteCredito'] = limiteCredito;
		jsonData['observacao'] = observacao;
		jsonData['tabelaPrecoModel'] = tabelaPrecoModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idPessoa = plutoRow.cells['idPessoa']?.value;
		idTabelaPreco = plutoRow.cells['idTabelaPreco']?.value;
		desde = Util.stringToDate(plutoRow.cells['desde']?.value);
		dataCadastro = Util.stringToDate(plutoRow.cells['dataCadastro']?.value);
		taxaDesconto = plutoRow.cells['taxaDesconto']?.value?.toDouble();
		limiteCredito = plutoRow.cells['limiteCredito']?.value?.toDouble();
		observacao = plutoRow.cells['observacao']?.value;
		tabelaPrecoModel = TabelaPrecoModel();
		tabelaPrecoModel?.nome = plutoRow.cells['tabelaPrecoModel']?.value;
	}	

	ClienteModel clone() {
		return ClienteModel(
			id: id,
			idPessoa: idPessoa,
			idTabelaPreco: idTabelaPreco,
			desde: desde,
			dataCadastro: dataCadastro,
			taxaDesconto: taxaDesconto,
			limiteCredito: limiteCredito,
			observacao: observacao,
		);			
	}

	
}

File: backend/cadastros/src/model/cliente_model.py
from src import db
from src.model.tabela_preco_model import TabelaPrecoModel


class ClienteModel(db.Model):
    __tablename__ = 'cliente'

    id = db.Column(db.Integer, primary_key=True)
    desde = db.Column(db.DateTime)
    data_cadastro = db.Column(db.DateTime)
    taxa_desconto = db.Column(db.Float)
    limite_credito = db.Column(db.Float)
    observacao = db.Column(db.String(250))
    id_pessoa = db.Column(db.Integer, db.ForeignKey('pessoa.id'), unique=True)
    id_tabela_preco = db.Column(db.Integer, db.ForeignKey('tabela_preco.id'))

    tabela_preco_model = db.relationship('TabelaPrecoModel', foreign_keys=[id_tabela_preco])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_pessoa = data.get('idPessoa')
        self.id_tabela_preco = data.get('idTabelaPreco')
        self.desde = data.get('desde')
        self.data_cadastro = data.get('dataCadastro')
        self.taxa_desconto = data.get('taxaDesconto')
        self.limite_credito = data.get('limiteCredito')
        self.observacao = data.get('observacao')

    def serialize(self):
        return {
            'id': self.id,
            'idPessoa': self.id_pessoa,
            'idTabelaPreco': self.id_tabela_preco,
            'desde': self.desde.isoformat(),
            'dataCadastro': self.data_cadastro.isoformat(),
            'taxaDesconto': self.taxa_desconto,
            'limiteCredito': self.limite_credito,
            'observacao': self.observacao,
            'tabelaPrecoModel': self.tabela_preco_model.serialize() if self.tabela_preco_model else None,
        }